lib/tapioca/compilers/dsl/active_record_columns.rb

```diff
@@ -0,0 +1,347 @@
+          add_method(
+            klass,
+            "#{column_name}_in_database",
+            methods_to_add,
+            return_type: getter_type
+          )
+          add_method(
+            klass,
+            "saved_change_to_#{column_name}",
+            methods_to_add,
+            return_type: "[#{getter_type}, #{getter_type}]"
+          )
+          add_method(
+            klass,
+            "saved_change_to_#{column_name}?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+          add_method(
+            klass,
+            "will_save_change_to_#{column_name}?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+
+          # Added by ActiveModel::Dirty
+          #
+          add_method(
+            klass,
+            "#{column_name}_change",
+            methods_to_add,
+            return_type: "[#{getter_type}, #{getter_type}]"
+          )
+          add_method(
+            klass,
+            "#{column_name}_changed?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+          add_method(
+            klass,
+            "#{column_name}_will_change!",
+            methods_to_add
+          )
+          add_method(
+            klass,
+            "#{column_name}_was",
+            methods_to_add,
+            return_type: getter_type
+          )
+          add_method(
+            klass,
+            "#{column_name}_previous_change",
+            methods_to_add,
+            return_type: "[#{getter_type}, #{getter_type}]"
+          )
+          add_method(
+            klass,
+            "#{column_name}_previously_changed?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+          add_method(
+            klass,
+            "#{column_name}_previously_was",
+            methods_to_add,
+            return_type: getter_type
+          )
+          add_method(
+            klass,
+            "restore_#{column_name}!",
+            methods_to_add
+          )
+
+          # Added by ActiveRecord::AttributeMethods::BeforeTypeCast
+          #
+          add_method(
+            klass,
+            "#{column_name}_before_type_cast",
+            methods_to_add,
+            return_type: "T.untyped"
+          )
+          add_method(
+            klass,
+            "#{column_name}_came_from_user?",
+            methods_to_add,
+            return_type: "T::Boolean"
+          )
+        end
+
+        sig do
+          params(
+            constant: T.class_of(ActiveRecord::Base),
+            column_name: String
+          ).returns([String, String])
+        end
+        def type_for(constant, column_name)
+          column_type = constant.attribute_types[column_name]
+
+          if Object.const_defined?(:StrongTypeGeneration) &&
+              !(constant.singleton_class < Object.const_get(:StrongTypeGeneration))
+            return ["T.untyped", "T.untyped"]
+          end
```

Comment:
I couldn't decide on the best way to break this conditional into 2 lines. Rubocop didn't auto fix it for me.

Reply:
It would be nice to extract that to a private method like:
```ruby
def generate_strong_types?
  Object.const_defined?(:StrongTypeGeneration) &&
  	constant.singleton_class < Object.const_get(:StrongTypeGeneration)
end
```